test/connectors/httpParsers/GiftAidSubmissionHttpParserSpec.scala

```diff
@@ -0,0 +1,99 @@
+ * distributed under the License is distributed on an "AS IS" BASIS,
+ * WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
+ * See the License for the specific language governing permissions and
+ * limitations under the License.
+ */
+
+package connectors.httpParsers
+
+import utils.UnitTest
+import connectors.httpParsers.GiftAidSubmissionHttpParser.CreateIncomeSourceHttpReads.read
+import models.{DesErrorBodyModel, DesErrorModel, GiftAidSubmissionResponseModel}
+import uk.gov.hmrc.http.HttpResponse
+import play.api.http.Status._
+import play.api.libs.json.Json
+
+class GiftAidSubmissionHttpParserSpec extends UnitTest {
+
+  def httpResponse(_status: Int, _body: String): HttpResponse = {
+    class SomeResponse(
+                        override val status: Int,
+                        override val body: String
+                      ) extends HttpResponse {
+      override def allHeaders: Map[String, Seq[String]] = Map()
+    }
+
+    new SomeResponse(_status, _body)
+  }
+
+  ".read" should {
+
+    "return a response model" when {
+
+      "the response contains an OK and the correct body" in {
+        val validBody = Json.prettyPrint(Json.obj("transactionReference" -> "some-transaction-id"))
+        val result = read("POST", "/some-url", httpResponse(OK, validBody))
+
+        result shouldBe Right(GiftAidSubmissionResponseModel("some-transaction-id"))
+      }
+
+    }
+
+    "a DES error" when {
+
+      "the response contains an OK, but a malformed body" in {
+        val invalidBody = Json.prettyPrint(Json.obj("malformed" -> "not-what-im-looking-for"))
+        val result = read("POST", "/some-url", httpResponse(OK, invalidBody))
+
+        result shouldBe Left(DesErrorModel(INTERNAL_SERVER_ERROR, DesErrorBodyModel.parsingError))
+      }
+
+      "the response status is INTERNAL_SERVER_ERROR" in {
+        val errorBody = Json.prettyPrint(Json.obj("code" -> "AWW_MAN", "reason" -> "not again"))
+        val result = read("POST", "/some-url", httpResponse(INTERNAL_SERVER_ERROR, errorBody))
+
+        result shouldBe Left(DesErrorModel(INTERNAL_SERVER_ERROR, DesErrorBodyModel("AWW_MAN", "not again")))
+      }
+
+      "the response status is SERVICE_UNAVAILABLE" in {
+        val errorBody = Json.prettyPrint(Json.obj("code" -> "AWW_MAN", "reason" -> "not again"))
+        val result = read("POST", "/some-url", httpResponse(SERVICE_UNAVAILABLE, errorBody))
+
+        result shouldBe Left(DesErrorModel(SERVICE_UNAVAILABLE, DesErrorBodyModel("AWW_MAN", "not again")))
+      }
+
+      "the response status is BAD_REQUEST" in {
+        val errorBody = Json.prettyPrint(Json.obj("code" -> "AWW_MAN", "reason" -> "not again"))
+        val result = read("POST", "/some-url", httpResponse(BAD_REQUEST, errorBody))
+
+        result shouldBe Left(DesErrorModel(BAD_REQUEST, DesErrorBodyModel("AWW_MAN", "not again")))
+      }
+
+      "the response status is FORBIDDEN" in {
```

Comment:
While I am a fan of these, I'm not sure if they should more accurately represent the actual errors we may be seeing